spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,319 @@
+    public boolean isResultNullable() {
+      return true;
+    }
+
+    @Override
+    public boolean isDeterministic() {
+      return true;
+    }
+  }
+
+  public static class TruncateInteger extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
+      }
+
+      return value - (((value % width) + width) % width);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.IntegerType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.IntegerType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "org.apache.iceberg.spark.functions.truncate(int,int)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
+      return invoke(input.getInt(0), input.getInt(1));
+    }
+  }
+
+  public static class TruncateLong extends TruncateBase<Long> {
+    // magic function for usage with codegen
+    public static long invoke(long value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
+      }
+
+      return value - (((value % width) + width) % width);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.LongType, DataTypes.IntegerType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.LongType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "org.apache.iceberg.spark.functions.truncate(long,int)";
+    }
+
+    @Override
+    public Long produceResult(InternalRow input) {
+      return invoke(input.getLong(0), input.getInt(1));
+    }
+  }
+
+  public static class TruncateString extends TruncateBase<String> {
+    // magic function for usage with codegen
+    public static UTF8String invoke(UTF8String value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
+      }
+
+      ByteBuffer bb = value.getByteBuffer();
+      CharSequence charSequence = StandardCharsets.UTF_8.decode(bb);
+      CharSequence truncated = UnicodeUtil.truncateString(charSequence, width);
+      ByteBuffer truncatedBytes = StandardCharsets.UTF_8.encode(CharBuffer.wrap(truncated));
+      return UTF8String.fromBytes(ByteBuffers.toByteArray(truncatedBytes));
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.StringType, DataTypes.IntegerType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.StringType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "org.apache.iceberg.spark.functions.truncate(string,int)";
```

Comment:
I'd probably make this `iceberg.truncate[width](string)`, but it's minor.

Reply:
Yeah I like that more. Updated.